File: moai-adk-ts/src/claude/hooks/workflow/file-monitor.ts
/**
 * @file file-monitor.ts
 * @description Unified file monitoring and checkpoint system for MoAI-ADK
 * @version 1.0.0
 * @tag @FEATURE:FILE-MONITORING-013
 */

import type { HookInput, HookResult, MoAIHook, FileEvent } from '../types';
import * as fs from 'fs';
import * as path from 'path';
import { EventEmitter } from 'events';

/**
 * File system event handler interface
 */
interface FileSystemEvent {
  type: 'created' | 'modified' | 'deleted';
  path: string;
}

/**
 * File Monitor Hook - TypeScript port of file_monitor.py
 */
export class FileMonitor extends EventEmitter implements MoAIHook {
  name = 'file-monitor';

  private projectRoot: string;
  private isRunning = false;
  private changedFiles = new Set<string>();
  private lastCheckpointTime = 0;
  private checkpointInterval = 300000; // 5 minutes in milliseconds
  private watcher?: fs.FSWatcher;

  // Essential file patterns to watch
  private watchPatterns = new Set([
    '.py',
    '.js',
    '.ts',
    '.md',
    '.json',
    '.yml',
    '.yaml',
  ]);

  // Directories to ignore
  private ignorePatterns = new Set([
    '.git',
    '__pycache__',
    'node_modules',
    '.pytest_cache',
    'dist',
    'build',
  ]);

  constructor(projectRoot?: string) {
    super();
    this.projectRoot = projectRoot || process.cwd();
  }

  async execute(input: HookInput): Promise<HookResult> {
    try {
      if (this.isMoAIProject()) {
        if (this.watchFiles()) {
          return {
            success: true,
            message: '📁 File monitoring started',
          };
        } else {
          return {
            success: true,
            message: '⚠️  Could not start file monitoring',
          };
        }
      }

      return { success: true };
    } catch (error) {
      // Silent failure to avoid breaking Claude Code session
      return { success: true };
    }
  }

  /**
   * Start file watching
   */
  watchFiles(): boolean {
    try {
      if (this.isRunning) {
        return true;
      }

      this.watcher = fs.watch(
        this.projectRoot,
        { recursive: true },
        (eventType, filename) => {
          if (filename) {
            const fullPath = path.join(this.projectRoot, filename);
            this.onFileChanged(fullPath);
          }
        }
      );

      this.isRunning = true;
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Stop file watching
   */
  stopWatching(): void {
    if (this.watcher && this.isRunning) {
      this.watcher.close();
      this.isRunning = false;
    }
  }

  /**
   * Handle file change event
   */
  private onFileChanged(filePath: string): void {
    if (!this.shouldMonitorFile(filePath)) {
      return;
    }

    this.changedFiles.add(filePath);

    // Emit event for external listeners
    const event: FileEvent = {
      path: filePath,
      type: 'modified', // Simplified for now
      timestamp: new Date(),
    };
    this.emit('fileChanged', event);

    // Check if should create checkpoint
    if (this.shouldCreateCheckpoint()) {
      this.createCheckpoint();
    }
  }

  /**
   * Determine if checkpoint should be created
   */
  private shouldCreateCheckpoint(): boolean {
    const currentTime = Date.now();

    // Create checkpoint if enough time has passed and files changed
    return (
      currentTime - this.lastCheckpointTime > this.checkpointInterval &&
      this.changedFiles.size > 0
    );
  }

  /**
   * Create checkpoint snapshot
   */
  private createCheckpoint(): boolean {
    try {
      const currentTime = Date.now();

      // Reset changed files and update timestamp
      this.changedFiles.clear();
      this.lastCheckpointTime = currentTime;

      // Emit checkpoint event
      this.emit('checkpoint', {
        timestamp: new Date(currentTime),
        changedFiles: Array.from(this.changedFiles),
      });

      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Check if file should be monitored
   */
  private shouldMonitorFile(filePath: string): boolean {
    const parsedPath = path.parse(filePath);
    const pathParts = filePath.split(path.sep);

    // Skip ignored directories
    for (const part of pathParts) {
      if (this.ignorePatterns.has(part)) {
        return false;
      }
    }

    // Only monitor files with relevant extensions
    return this.watchPatterns.has(parsedPath.ext);
  }

  /**
   * Check if this is a MoAI project
   */
  private isMoAIProject(): boolean {
    const moaiPath = path.join(this.projectRoot, '.moai');
    return fs.existsSync(moaiPath);
  }

  /**
   * Get list of changed files since last checkpoint
   */
  getChangedFiles(): string[] {
    return Array.from(this.changedFiles);
  }

  /**
   * Get monitoring statistics
   */
  getStats(): {
    isRunning: boolean;
    changedFiles: number;
    lastCheckpoint: Date | null;
  } {
    return {
      isRunning: this.isRunning,
      changedFiles: this.changedFiles.size,
      lastCheckpoint:
        this.lastCheckpointTime > 0 ? new Date(this.lastCheckpointTime) : null,
    };
  }
}

/**
 * CLI entry point for Claude Code compatibility
 */
export async function main(): Promise<void> {
  try {
    const projectRoot = process.cwd();
    const monitor = new FileMonitor(projectRoot);

    const result = await monitor.execute({});

    if (result.message) {
      console.log(result.message);
    }

    // Keep process alive for monitoring
    if (monitor.getStats().isRunning) {
      process.on('SIGINT', () => {
        monitor.stopWatching();
        process.exit(0);
      });

      process.on('SIGTERM', () => {
        monitor.stopWatching();
        process.exit(0);
      });
    }
  } catch (error) {
    // Silent failure to avoid breaking Claude Code session
  }
}

// Execute if run directly
if (require.main === module) {
  main().catch(() => {
    // Silent failure
  });
}

File: .claude/hooks/moai/file_monitor.py
#!/usr/bin/env python3
"""
Unified File Monitor for MoAI-ADK - Optimized v0.2.0
Combines file watching and auto checkpoint functionality.

@REQ:FILE-MONITOR-001
@FEATURE:FILE-MONITORING-OPT
@TEST:UNIT-FILE-MONITOR-SIZE
"""

import os
import time
from pathlib import Path

try:
    from watchdog.events import FileSystemEventHandler
    from watchdog.observers import Observer

    WATCHDOG_AVAILABLE = True
except ImportError:
    WATCHDOG_AVAILABLE = False


class FileMonitor:
    """Unified file monitoring and checkpoint system

    @FEATURE:FILE-MONITORING-OPT
    Merged from file_watcher.py (323 lines) + auto_checkpoint.py (222 lines)
    Reduced to ~150 lines while preserving core functionality:
    - File change detection
    - Auto checkpoint creation
    """

    def __init__(self, project_root: Path):
        self.project_root = project_root
        self.observer = None
        self.is_running = False
        self.changed_files: set[str] = set()
        self.last_checkpoint_time = 0
        self.checkpoint_interval = 300  # 5 minutes

        # Essential file patterns to watch
        self.watch_patterns = {".py", ".js", ".ts", ".md", ".json", ".yml", ".yaml"}

        # Directories to ignore
        self.ignore_patterns = {".git", "__pycache__", "node_modules", ".pytest_cache"}

    def watch_files(self) -> bool:
        """Start file watching

        @FEATURE:FILE-CHANGE-DETECTION
        Essential functionality for file monitoring
        """
        if not WATCHDOG_AVAILABLE:
            return False

        try:
            self.observer = Observer()
            handler = MoAIFileHandler(self)
            self.observer.schedule(handler, str(self.project_root), recursive=True)
            self.observer.start()
            self.is_running = True
            return True
        except Exception:
            return False

    def stop_watching(self):
        """Stop file watching"""
        if self.observer and self.is_running:
            self.observer.stop()
            self.observer.join()
            self.is_running = False

    def on_file_changed(self, file_path: str):
        """Handle file change event

        @FEATURE:FILE-CHANGE-HANDLER
        Essential for responding to file changes
        """
        file_path_obj = Path(file_path)

        # Check if file should be monitored
        if not self._should_monitor_file(file_path_obj):
            return

        self.changed_files.add(file_path)

        # Check if should create checkpoint
        if self.should_create_checkpoint():
            self.create_checkpoint()

    def should_create_checkpoint(self) -> bool:
        """Determine if checkpoint should be created

        @FEATURE:CHECKPOINT-LOGIC
        Essential for auto checkpoint functionality
        """
        current_time = time.time()

        # Create checkpoint if enough time has passed and files changed
        if (
            current_time - self.last_checkpoint_time > self.checkpoint_interval
            and self.changed_files
        ):
            return True

        return False

    def create_checkpoint(self) -> bool:
        """Create checkpoint snapshot

        @FEATURE:AUTO-CHECKPOINT
        Essential for Git checkpoint functionality
        """
        try:
            # Simple checkpoint creation
            current_time = time.time()

            # Reset changed files and update timestamp
            self.changed_files.clear()
            self.last_checkpoint_time = current_time

            # In a real implementation, this would create a Git checkpoint
            # For now, just return success
            return True

        except Exception:
            return False

    def _should_monitor_file(self, file_path: Path) -> bool:
        """Check if file should be monitored"""
        # Skip ignored directories
        for part in file_path.parts:
            if part in self.ignore_patterns:
                return False

        # Only monitor files with relevant extensions
        if file_path.suffix in self.watch_patterns:
            return True

        return False


class MoAIFileHandler(FileSystemEventHandler):
    """File system event handler for MoAI file monitor"""

    def __init__(self, monitor: FileMonitor):
        super().__init__()
        self.monitor = monitor

    def on_modified(self, event):
        """Handle file modification events"""
        if not event.is_directory:
            self.monitor.on_file_changed(event.src_path)

    def on_created(self, event):
        """Handle file creation events"""
        if not event.is_directory:
            self.monitor.on_file_changed(event.src_path)


def main():
    """Main entry point for file monitoring"""
    try:
        project_root = Path(os.getcwd())
        monitor = FileMonitor(project_root)

        # Start monitoring if in MoAI project
        if (project_root / ".moai").exists():
            if monitor.watch_files():
                print("📁 File monitoring started")
            else:
                print("⚠️  Could not start file monitoring")

    except Exception:
        # Silent failure to avoid breaking Claude Code session
        pass


if __name__ == "__main__":
    main()
